lib/src/rules/weak_mode_unnecessary_null_checks.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Unnecessary null check for non-nullable value.';
+
+const _details = r'''
+
+Unnecessary null check for non-nullable value in a opted-out library.
+
+**BAD:**
+```
+nonNullable?.property;
+if (nonNullable != null) {}
+```
+
+**GOOD:**
+```
+nonNullable.property;
+```
+
+''';
+
+class WeakModeUnnecessaryNullChecks extends LintRule implements NodeLintRule {
+  WeakModeUnnecessaryNullChecks()
+      : super(
+            name: 'weak_mode_unnecessary_null_checks',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends RecursiveAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
+      return;
+    }
+    super.visitCompilationUnit(node);
+  }
+
+  @override
+  void visitPropertyAccess(PropertyAccess node) {
+    if (node.operator.type == TokenType.QUESTION_PERIOD &&
+        context.typeSystem.isNonNullable(node.target.staticType)) {
```

Comment:
@bwilkerson I don't know why but the type of an element in opted-in library is always with the * suffix and so `context.typeSystem.isNonNullable(node.target.staticType)` is never true. I thought it will be `String` instead of `String*`. Am I missing something?

Reply:
As per the semantics from the spec, in an opted out library all types (whether they come from opted-in or opted-out libraries) are converted to have a * suffix.

Reply:
Thanks for the explanation. I thought * was only used for types coming from opted-out into opted-in.